### PR TITLE
Auction: implementation

### PR DESCRIPTION
**Pull request by gangov**

### CODE REVIEW

File: contracts/auctions/src/contract.rs [hunk 1]
```diff
@@ -0,0 +1,430 @@
...
+            }
+        }
+    }
+
+    #[allow(dead_code)]
+    pub fn finalize_auction(env: Env, auction_id: u64) -> Result<(), ContractError> {
+        let mut auction = get_auction_by_id(&env, auction_id)?;
+        auction.seller.require_auth();
+
+        let token_client = token::Client::new(&env, &auction.currency);
```
Comment (ueco-jb) [inline contracts/auctions/src/contract.rs:183]: nit: Token client is needed lower. No need to initialize it here.

File: contracts/auctions/src/contract.rs [hunk 2]
```diff
@@ -0,0 +1,430 @@
...
+                save_auction_by_id(&env, auction_id, &auction)?;
+                save_auction_by_seller(&env, &auction.seller, &auction)?;
+
+                Ok(())
+            }
+        }
+    }
+
+    #[allow(dead_code)]
+    pub fn finalize_auction(env: Env, auction_id: u64) -> Result<(), ContractError> {
```
Comment (ueco-jb) [inline contracts/auctions/src/contract.rs:179]: I think this function needs some refactoring. 
There are 3 cases here:
- min price not reached
- no one bidded
- the correct case

Right now it's this weird thing where you repeat 3x in code the same sections
```
           auction.status = AuctionStatus::Ended;

            save_auction_by_id(&env, auction_id, &auction)?;
            save_auction_by_seller(&env, &auction.seller, &auction)?;
            update_auction(&env, auction_id, auction)?;

            return Ok(());
```
which is the redundancy of the highest order. Plus it makes code harder to read.

Make it into a big if-else of match statement and make the common operations happend before and after that statement.

File: contracts/auctions/src/contract.rs [hunk 3]
```diff
@@ -0,0 +1,430 @@
...
+        let mut auction = get_auction_by_id(&env, auction_id)?;
+        auction.seller.require_auth();
+
+        if auction.status != AuctionStatus::Active {
+            log!(env, "Auction: Pause: Cannot pause inactive/ended auction.");
+            return Err(ContractError::AuctionNotActive);
+        }
+
+        if env.ledger().timestamp() > auction.end_time {
+            log!(env, "Auction: Buy now: Auction expired");
```
Comment (ueco-jb) [inline contracts/auctions/src/contract.rs:344]:
```suggestion
            log!(env, "Auction: Pause: Auction expired");
```

File: contracts/auctions/src/contract.rs [hunk 4]
```diff
@@ -0,0 +1,430 @@
...
+        let mut auction = get_auction_by_id(env, auction_id)?;
+        auction.seller.require_auth();
+
+        if auction.status != AuctionStatus::Paused {
+            log!(env, "Auction: Pause: Cannot activate unpaused auction.");
+            return Err(ContractError::AuctionNotPaused);
+        }
+
+        if env.ledger().timestamp() > auction.end_time {
+            log!(env, "Auction: Buy now: Auction expired");
```
Comment (ueco-jb) [inline contracts/auctions/src/contract.rs:366]:
```suggestion
            log!(env, "Auction: Unpause: Auction expired");
```

File: contracts/auctions/src/contract.rs
Comment (ueco-jb) [inline contracts/auctions/src/contract.rs:1]: Implement update message (contract upgrade).

Comment (ueco-jb) [inline contracts/auctions/src/contract.rs:1]: Add events to each transaction, I saw only one or two singular events.

**Review by ueco-jb — COMMENTED** [review]
Initial review of the implementation alone.

File: contracts/auctions/src/lib.rs [hunk 5]
```diff
@@ -1,84 +1,21 @@
...
-    Cancelled,
+mod contract;
+mod error;
+mod storage;
+
+#[cfg(test)]
+mod test;
+
+pub mod collection {
+    type NftId = u64;
```
Comment (ueco-jb) [inline contracts/auctions/src/lib.rs:11]: Isn't it possible to import it through the hash? I think I did that once.

File: contracts/auctions/src/storage.rs [hunk 6]
```diff
@@ -0,0 +1,290 @@
...
+        .get::<_, u64>(&DataKey::AuctionId)
+        .unwrap_or_default()
+        + 1u64;
+    env.storage().instance().set(&DataKey::AuctionId, &id);
+    env.storage()
+        .instance()
+        .extend_ttl(LIFETIME_THRESHOLD, BUMP_AMOUNT);
+
+    Ok(id)
+}
```
Comment (ueco-jb) [inline contracts/auctions/src/storage.rs:76]: Good use of the instance storage. :+1:

File: contracts/auctions/src/storage.rs [hunk 7]
```diff
@@ -0,0 +1,290 @@
...
+) -> Result<Vec<Auction>, ContractError> {
+    let start_index = start_index.unwrap_or(DEFAULT_INDEX);
+    let limit = limit.unwrap_or(DEFAULT_LIMIT);
+
+    let mut auctions = vec![&env];
+
+    for id in start_index..=limit {
+        match get_auction_by_id(env, id as u64) {
+            Ok(auction) => auctions.push_back(auction),
+            Err(ContractError::AuctionNotFound) => break,
```
Comment (ueco-jb) [inline contracts/auctions/src/storage.rs:91]:
```suggestion
            Err(ContractError::AuctionNotFound) => continue,
```
Why stopping the whole loop? You are not returning the error anyway.

File: contracts/auctions/src/storage.rs [hunk 8]
```diff
@@ -0,0 +1,290 @@
...
+    env.storage()
+        .persistent()
+        .has(&DataKey::HighestBid(auction_id))
+        .then(|| {
+            env.storage().persistent().extend_ttl(
+                &DataKey::HighestBid(auction_id),
+                LIFETIME_THRESHOLD,
+                BUMP_AMOUNT,
+            )
+        });
```
Comment (ueco-jb) [inline contracts/auctions/src/storage.rs:248]: All bid related stuff should also be in an instance storage I think.

File: contracts/auctions/src/storage.rs [hunk 9]
```diff
@@ -0,0 +1,290 @@
...
+        .instance()
+        .extend_ttl(LIFETIME_THRESHOLD, BUMP_AMOUNT);
+
+    Ok(())
+}
+
+pub fn validate_input_params(env: &Env, values_to_check: &[&u64]) -> Result<(), ContractError> {
+    values_to_check.iter().for_each(|i| {
+        if i < &&1 {
+            log!(&env, "Auction: Validate input: Invalid inputs used");
```
Comment (ueco-jb) [inline contracts/auctions/src/storage.rs:181]: Since you are checking the specific case, you don't need to be so general.
```suggestion
            log!(&env, "Auction: Validate input: parameters can not be less then 1");
```

File: contracts/auctions/src/contract.rs [hunk 10]
```diff
@@ -0,0 +1,430 @@
...
+    error::ContractError,
+    storage::{
+        generate_auction_id, get_admin, get_auction_by_id, get_auctions, get_auctions_by_seller_id,
+        get_highest_bid, is_initialized, save_admin, save_auction_by_id, save_auction_by_seller,
+        set_highest_bid, set_initialized, update_admin, update_auction, validate_input_params,
+        Auction, AuctionStatus, ItemInfo,
+    },
+    token,
+};
+const ONE: u64 = 1u64;
```
[the remark this answers is not in the document]
Comment (gangov) [reply contracts/auctions/src/contract.rs:14]: looks weird, I know, I don't want to hardcode 1 and forget about it - reason for that is that currently the marketplace allows to handle only auctions for a single item, not multiple items.

Let's say a seller has 5 items, on the marketplace they can sell just one 1, hence `const ONE`. 

I guess the next improvement can be allowing to sell entire collections, among the other type of auctions

File: contracts/token/Cargo.toml
Comment (ueco-jb) [inline contracts/token/Cargo.toml:1]: On a side - token should be added in a separate PR.

**Review by ueco-jb — COMMENTED** [review]
Some more comments regarding the storage, I will review the tests after those are solved.

File: contracts/auctions/src/contract.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (ueco-jb) [reply contracts/auctions/src/contract.rs:14]: > Let's say a seller has 5 items, on the marketplace they can sell just one 1, hence const ONE.

I still don't get what's better in getting a constant instead of a number that literally represent that word?

[the remark this answers is not in the document]
Comment (gangov) [reply contracts/auctions/src/contract.rs]: removed

File: contracts/auctions/src/contract.rs [hunk 11]
```diff
@@ -0,0 +1,430 @@
...
+        let auction = Auction {
+            id,
+            item_info,
+            seller: seller.clone(),
+            highest_bid: None,
+            // we use the seller's address as we cannot add `Option<Address>` in the struct
+            highest_bidder: seller.clone(),
+            end_time,
+            status: AuctionStatus::Active,
+            currency,
```
[the remark this answers is not in the document]
Comment (gangov) [reply contracts/auctions/src/contract.rs:79]: exactly that, changing it

File: contracts/auctions/src/contract.rs [hunk 12]
```diff
@@ -0,0 +1,430 @@
...
+                    &(bid_amount as i128),
+                );
+
+                set_highest_bid(&env, auction_id, bid_amount, bidder.clone())?;
+
+                // Update auction state
+                //TODO: in case we have separate storage isn't the two below
+                //redundant?
+                auction.highest_bid = Some(bid_amount);
+                auction.highest_bidder = bidder;
```
[the remark this answers is not in the document]
Comment (gangov) [reply contracts/auctions/src/contract.rs:145]: what about `highest_bid` that is also kept in a separate storage, imo we can remove both `highest_bid` and `highest_bidder` from `Auction`

File: contracts/auctions/src/contract.rs (hunk 1 above)
Comment (gangov) [reply contracts/auctions/src/contract.rs]: moved the client after the assertions

File: contracts/auctions/src/contract.rs (hunk 2 above)
Comment (gangov) [reply contracts/auctions/src/contract.rs:181]: split the logic and placed it in `utils.rs` now it should be more readable

File: contracts/auctions/src/contract.rs (hunk 4 above)
Comment (gangov) [reply contracts/auctions/src/contract.rs]: fixed a few others as well in that line of thoughts

File: contracts/auctions/src/contract.rs (hunk 3 above)
Comment (gangov) [reply contracts/auctions/src/contract.rs]: fixed a few others in that lines of thought

File: contracts/auctions/src/contract.rs (hunk 12 above)
[the remark this answers is not in the document]
Comment (ueco-jb) [reply contracts/auctions/src/contract.rs]: Leave `highest_bid` because it's accessible in the same structure as the rest of the data necessary for the auction to operate.

File: contracts/auctions/src/lib.rs (hunk 5 above)
Comment (gangov) [reply contracts/auctions/src/lib.rs:12]: couldn't find any other examples that use custom type from a different contract.

The complier needs to satisfy this:

```
cannot find type `NftId` in this scope
not found in this scope 
```

File: contracts/auctions/src/contract.rs [hunk 13]
```diff
@@ -0,0 +1,443 @@
...
+
+        let id = generate_auction_id(&env)?;
+        let end_time = env.ledger().timestamp() + duration;
+
+        let auction = Auction {
+            id,
+            item_info,
+            seller: seller.clone(),
+            highest_bid: None,
+            // we use the seller's address as we cannot add `Option<Address>` in the struct
```
Comment (ueco-jb) [inline contracts/auctions/src/contract.rs:89]:
```suggestion
```

File: contracts/auctions/src/contract.rs [hunk 14]
```diff
@@ -0,0 +1,443 @@
...
+
+                set_highest_bid(&env, auction_id, bid_amount, bidder.clone())?;
+
+                // Update auction state
+                auction.highest_bid = Some(bid_amount);
+                update_auction(&env, auction_id, auction.clone())?;
+                save_auction_by_id(&env, auction_id, &auction)?;
+                save_auction_by_seller(&env, &auction.seller, &auction)?;
+            }
+        }
```
Comment (ueco-jb) [inline contracts/auctions/src/contract.rs:190]:
```suggestion
        match auction.highest_bid {
            Some(current_highest_bid) if bid_amount > current_highest_bid => {
                // refund the previous highest bidder
                let old_bid_info = get_highest_bid(&env, auction_id)?;

                token_client.transfer(
                    &env.current_contract_address(),
                    &old_bid_info.bidder,
                    &(old_bid_info.bid as i128),
                );
            }
            Some(_) => {
                log!(
                    &env,
                    "Auction: Place Bid: Bid not enough. Amount bid: ",
                    bid_amount
                );
                return Err(ContractError::BidNotEnough);
            }
            None => { }
        };
        token_client.transfer(
            &bidder,
            &env.current_contract_address(),
            &(bid_amount as i128),
        );

        set_highest_bid(&env, auction_id, bid_amount, bidder.clone())?;

        // Update auction state
        auction.highest_bid = Some(bid_amount);
        update_auction(&env, auction_id, auction.clone())?;
        save_auction_by_id(&env, auction_id, &auction)?;
        save_auction_by_seller(&env, &auction.seller, &auction)?;
```

File: contracts/auctions/src/utils.rs [hunk 15]
```diff
@@ -0,0 +1,109 @@
...
+    !auction
+        .item_info
+        .minimum_price
+        .and_then(|min_price| {
+            auction
+                .highest_bid
+                .map(|highest_bid| min_price > highest_bid)
+        })
+        .unwrap_or(false)
+}
```
Comment (ueco-jb) [inline contracts/auctions/src/utils.rs:44]: This is confusing. You are checking if `minimum_price_reached` while really you check if the minimum price has not been reached.

```suggestion
pub(crate) fn minimum_price_reached(auction: &Auction) -> bool {
    auction
        .item_info
        .minimum_price
        .map_or(true, |min_price| {
            auction
                .highest_bid
                .map_or(false, |highest_bid| highest_bid >= min_price)
        })
}
```

- if there's no minimum_price, we assume the minimum price has been reached (since there's no minimum).
- if there's no highest_bid, the minimum price has not been reached.
- the final comparison checks whether the highest_bid is greater than or equal to the minimum_price.

File: contracts/auctions/src/utils.rs [hunk 16]
```diff
@@ -0,0 +1,109 @@
...
+        &(highest_bid.bid as i128),
+    );
+    auction.status = AuctionStatus::Ended;
+    save_auction(env, auction_id, auction)?;
+    log!(
+        env,
+        "Auction: Finalize auction: Miniminal price not reached"
+    );
+    Ok(())
+}
```
Comment (ueco-jb) [inline contracts/auctions/src/utils.rs:65]: You created a helper which obfuscates the logic and is used only once. Remove it and put that functionality back to the code.

File: contracts/auctions/src/utils.rs [hunk 17]
```diff
@@ -0,0 +1,109 @@
...
+
+pub(crate) fn end_auction_without_bids(
+    env: &Env,
+    auction: &mut Auction,
+    auction_id: u64,
+) -> Result<(), ContractError> {
+    auction.status = AuctionStatus::Ended;
+    save_auction(env, auction_id, auction)?;
+    Ok(())
+}
```
Comment (ueco-jb) [inline contracts/auctions/src/utils.rs:75]: What's the point of that? Remove that.

File: contracts/auctions/src/storage.rs [hunk 18]
```diff
@@ -0,0 +1,319 @@
...
+        log!(&env, "Auction update auction: Id missmatch");
+        panic_with_error!(&env, ContractError::IDMissmatch);
+    }
+    env.storage().instance().set(&auction.id, &auction);
+    env.storage()
+        .instance()
+        .extend_ttl(LIFETIME_THRESHOLD, BUMP_AMOUNT);
+
+    Ok(())
+}
```
Comment (ueco-jb) [inline contracts/auctions/src/storage.rs:176]: What's the point of this helper if it does save the same thing as `save_auction_by_id`?

**Review by ueco-jb — COMMENTED** [review]
Pushing the review

File: contracts/auctions/src/utils.rs [hunk 19]
```diff
@@ -0,0 +1,35 @@
...
+    }
+    if env.ledger().timestamp() < auction.end_time {
+        log!(
+            env,
+            "Auction: Finalize auction: Auction cannot be ended early"
+        );
+        return Err(ContractError::AuctionNotFinished);
+    }
+    Ok(())
+}
```
Comment (ueco-jb) [inline contracts/auctions/src/utils.rs:27]: Again, a helper used in only once place that hides actual conditions being checked.

File: contracts/auctions/src/contract.rs [hunk 20]
```diff
@@ -0,0 +1,433 @@
...
+            );
+            return Err(ContractError::AuctionNotActive);
+        }
+
+        if env.ledger().timestamp() > auction.end_time {
+            log!(&env, "Auction: Pause: Auction expired: ", auction_id);
+            return Err(ContractError::AuctionNotActive);
+        }
+
+        auction.status = AuctionStatus::Paused;
```
Comment (ueco-jb) [inline contracts/auctions/src/contract.rs:322]: `AuctionStatus::Paused` should be checked during bidding, finalizing and buy now. This is not tested at all.

File: contracts/auctions/src/contract.rs [hunk 21]
```diff
@@ -0,0 +1,433 @@
...
+
+        save_auction_by_id(&env, auction_id, &auction)?;
```
Comment (ueco-jb) [inline contracts/auctions/src/contract.rs:324]: Use helpers you have implemented.

```suggestion
        save_auction(&env, auction_id, &auction)?;
```

File: contracts/auctions/src/contract.rs [hunk 22]
```diff
@@ -0,0 +1,433 @@
...
+        }
+
+        if env.ledger().timestamp() > auction.end_time {
+            log!(env, "Auction: Unpause: Auction expired: ", auction_id);
+            return Err(ContractError::AuctionNotActive);
+        }
+
+        auction.status = AuctionStatus::Active;
+
+        save_auction_by_id(env, auction_id, &auction)?;
```
Comment (ueco-jb) [inline contracts/auctions/src/contract.rs:352]:
```suggestion
        save_auction(env, auction_id, &auction)?;
```

**Review by ueco-jb — COMMENTED** [review]
Some more comments.

File: contracts/auctions/src/contract.rs [hunk 23]
```diff
@@ -0,0 +1,433 @@
...
+        let old_highest_bid = get_highest_bid(&env, auction_id)?;
+
+        // refund any previous highest bid
+        let token = token::Client::new(&env, &auction.currency);
+
+        token.transfer(
+            &env.current_contract_address(),
+            &old_highest_bid.bidder,
+            &(old_highest_bid.bid as i128),
+        );
```
[the remark this answers is not in the document]
Comment (gangov) [reply contracts/auctions/src/contract.rs]: currently we have 
```Rust
let old_highest_bid = get_highest_bid(&env, auction_id)?;
```
the return is of type `HighestBidder` which fields _aren't_ `Option`

I've added 2 more tests cases to better express the two different scenarios 
`buy_now_should_refund_previous_buyer` and `buy_now_should_work_when_no_previous_bid`

I'll put the refund in a if statement for when there's an some amount to be refunded

File: contracts/auctions/src/contract.rs (hunk 20 above)
Comment (gangov) [reply contracts/auctions/src/contract.rs:344]: that's checked inside the following tests
`buy_now` is tested when auction is paused inside `buy_now_should_fail_when_status_is_different_from_active`
`finalize_auction` is tested when the auction is paused in `fail_to_finalize_auction_when_not_correct_state`
`place_bid` is _now_ tested when auction is paused inside `unpause_changes_status_and_second_attempt_fails_to_unpause` which I also extended with all 2 previous ones.

File: contracts/auctions/src/contract.rs [hunk 24]
```diff
@@ -0,0 +1,455 @@
...
+                &(highest_bid.bid as i128),
+            );
+
+            let nft_client = collection::Client::new(&env, &auction.item_info.collection_addr);
+            nft_client.safe_transfer_from(
+                &env.current_contract_address(),
+                &auction.seller,
+                &highest_bid.bidder,
+                &auction.item_info.item_id,
+                &1,
```
Comment (ueco-jb) [inline contracts/auctions/src/contract.rs:216]: I guess the amount of items should also be specified in the `ItemInfo` structure?

**Review by ueco-jb — APPROVED** [review]
Let's merge it and do the comment in the follow up.

File: contracts/auctions/src/contract.rs (hunk 24 above)
Comment (gangov) [reply contracts/auctions/src/contract.rs:216]: moving this to a separate issue
